rust/otap-dataflow/crates/engine/src/shared/message.rs

```diff
@@ -3,93 +3,316 @@
+        receiver: flume::Receiver<T>,
+        pipeline_ctx: &PipelineContext,
+        channel_metrics: &mut ChannelMetricsRegistry,
+        channel_id: Cow<'static, str>,
+        channel_kind: &'static str,
+        capacity: u64,
+    ) -> Self {
+        let attrs = pipeline_ctx.channel_attribute_set(
+            channel_id,
+            channel_kind,
+            CHANNEL_MODE_SHARED,
+            CHANNEL_TYPE_MPMC,
+            CHANNEL_IMPL_FLUME,
+        );
+        let metrics = pipeline_ctx
+            .metrics_registry()
+            .register::<ChannelReceiverMetrics>(attrs);
+        let handle = Arc::new(Mutex::new(ChannelReceiverMetricsState::new(
+            metrics, capacity,
+        )));
+        channel_metrics.register(ChannelMetricsHandle::SharedReceiver(handle.clone()));
+        let mut receiver = Self::mpmc(receiver);
+        receiver.metrics = Some(handle);
+        receiver
+    }
+
+    pub(crate) fn into_mpsc(self) -> Result<tokio::sync::mpsc::Receiver<T>, Self> {
+        let SharedReceiver { inner, metrics } = self;
+        match inner {
+            SharedReceiverInner::Mpsc(receiver) => Ok(receiver),
+            SharedReceiverInner::Mpmc(receiver) => Err(Self {
+                inner: SharedReceiverInner::Mpmc(receiver),
+                metrics,
+            }),
+        }
+    }
+
     /// Receives a message from the channel.
     pub async fn recv(&mut self) -> Result<T, RecvError> {
-        match self {
-            SharedReceiver::MpscReceiver(receiver) => {
-                receiver.recv().await.ok_or(RecvError::Closed)
-            }
-            SharedReceiver::MpmcReceiver(receiver) => {
-                receiver.recv().map_err(|_| RecvError::Closed)
+        let result = match &mut self.inner {
+            SharedReceiverInner::Mpsc(receiver) => receiver.recv().await.ok_or(RecvError::Closed),
+            SharedReceiverInner::Mpmc(receiver) => receiver.recv().map_err(|_| RecvError::Closed),
+        };
+
+        if let Some(metrics) = &self.metrics {
+            if let Ok(mut metrics) = metrics.lock() {
+                match &result {
+                    Ok(_) => metrics.record_recv_ok(),
+                    Err(RecvError::Empty) => metrics.record_recv_error_empty(),
+                    Err(RecvError::Closed) => metrics.record_recv_error_closed(),
+                }
             }
         }
+
+        result
     }
 
     /// Tries to receive a message from the channel.
     pub fn try_recv(&mut self) -> Result<T, RecvError> {
-        match self {
-            SharedReceiver::MpscReceiver(receiver) => {
+        let result = match &mut self.inner {
+            SharedReceiverInner::Mpsc(receiver) => {
                 receiver.try_recv().map_err(|_| RecvError::Closed)
             }
-            SharedReceiver::MpmcReceiver(receiver) => {
+            SharedReceiverInner::Mpmc(receiver) => {
                 receiver.try_recv().map_err(|_| RecvError::Closed)
             }
+        };
```

Comment:
not directly related to this PR - but it seems we are mapping all errors to `RecvError::Closed`. Better to check for `TryRecvError::Empty` and `TryRecvError::Disconnected` (for both tokio and flume) and handle accordingly.